### DISCUSSION TITLE
Why use channel for concurrency control? mutex seems better.

### DISCUSSION THREAD

**Discussion by aa624545345** [moby/moby#46306] [General]
In sharedtemp.go, channel stfcState is used for concurrency control. 
```go
type sharedTempFileConverter struct {
	...
	st   chan stfcState
}
```

This channel is similar to using a mutex. 
```go
func (c *sharedTempFileConverter) openNew(st stfcState, f *os.File, stat os.FileInfo) (*sharedFileReader, error) {
	......
	c.st <- st // Lock

	dst, size, convErr := c.convert(f)

	st = <-c.st // Unlock
        ......
}
```

Why not use mutex for concurrency control? In my tests, Mutex performed better than channel.
```go
// mutex.go
func main() {
	var mu sync.Mutex
	count := 0
	for i := 0; i < 1000000; i++ {
		go func() {
			mu.Lock()
			count++
			mu.Unlock()
		}()
	}
	for count < 1000000 {
	}
	fmt.Println("Count:", count)
}
// [root@localhost]# time ./mutex
// Count: 1000000

// real	0m1.289s
// user	0m3.156s
// sys	0m0.473s

// channel.go
func main() {
	mutex := NewMutex()
	count := 0
	for i := 0; i < 1000000; i++ {
		go func() {
			mutex.Lock()
			count++
			mutex.Unlock()
		}()
	}
	for count < 1000000 {
	}
	fmt.Println("Count:", count)
}
// [root@localhost]# time ./channel
// Count: 1000000

// real	0m5.844s
// user	0m20.418s
// sys	0m8.729s
```

**Comment by corhere** [comment]
You have the lock and unlock operations swapped around.
```
func (c *sharedTempFileConverter) openNew(st stfcState, f *os.File, stat os.FileInfo) (*sharedFileReader, error) {
	......
	c.st <- st // Unlock (send state to channel)

	dst, size, convErr := c.convert(f)

	st = <-c.st // Lock (receive state from channel)
        ......
}
```

[As I explained in the commit message:](https://github.com/moby/moby/pull/43294/commits/6d5bc071890a02fb2f4482e73e7cb4071a3ce370)
> A channel is used for concurrency control so that the lock can be
> acquired inside a select statement. While not currently utilized, this
> makes it possible to add support for cancellation to
> sharedTempFileConverter in the future.

I also followed the same pattern in `LogFile`, which has a comment explaining the other considerations which went into this design choice.

https://github.com/moby/moby/blob/a65c948e7edf99eeb3c1170342fe86932ce9f9eb/daemon/logger/loggerutils/logfile.go#L50-L64

The excellent [Rethinking Classical Concurrency Patterns](https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view) talk goes into much greater depth on the subject. I highly recommend reading ([or watching](https://www.youtube.com/watch?v=5zXAHh5tJqQ)) it. It was my main source of inspiration for using channels when I overhauled the logging subsystem.

While mutexes _may be_ faster than channels in your particular microbenchmark today, that does not make them inherently "better" in all respects. Speed is not the only dimension to consider. Concurrent code with shared memory is very hard to get right. Case in point: your benchmark code contains a race condition. And so did the old logger code which used mutexes. Race conditions lead to wrong answers, and wrong answers are almost never useful. The fastest program known is one which exits immediately, but unless you are writing `/bin/false`, such a program would provide no value to you. I think it would be uncontroversial to say that in general, if there is a choice between fast and wrong, or slow and correct, slow and correct would be the better choice. Sharing memory through channels makes it easier to write _correct_ concurrent code, at the cost of a slight performance penalty...maybe. Compiler and runtime improvements will close the performance gap over time. And as noted in the description of #43294, switching over to channel-based code actually slightly _increased_ performance over the implementation which used mutexes and condition variables. The time it takes to acquire or release a lock is dwarfed by the time to perform I/O and everything else the logging subsystem has to handle.

tl;dr channels are used for concurrency control because they make it easier to write correct concurrent code, without impacting performance nearly as much as microbenchmarks may suggest.
